NIFI-5752: Load balancing fails with wildcard certs

kotarot opened the pull request:
Thank you for submitting a contribution to Apache NiFi.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [X] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [X] Does your PR title start with NIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [X] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [X] Is your initial contribution a single, squashed commit?

### For code changes:
- [X] Have you ensured that the full suite of tests is executed via mvn -Pcontrib-check clean install at the root nifi folder?
- [ ] Have you written or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the LICENSE file, including the main LICENSE file under nifi-assembly?
- [ ] If applicable, have you updated the NOTICE file, including the main NOTICE file found under nifi-assembly?
- [ ] If adding new Properties, have you added .displayName in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

ijokarumawak commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/queue/clustered/server/ClusterLoadBalanceAuthorizer.java:
```diff
@@ -40,28 +42,23 @@ public ClusterLoadBalanceAuthorizer(final ClusterCoordinator clusterCoordinator,
+            final HostnameVerifier verifier = new DefaultHostnameVerifier();
+            if (verifier.verify(nodeId, sslSession)) {
+                logger.debug("Authorizing Client to Load Balance data");
                 return;
```
By #3109, we need to return the client peer description when authorization passes. For the best informative result for data provenance, we need to do:
- If any SAN exists in the known nodeIds, then return the matched SAN (this can be done by the existing code), this way, we can identify which node sent the request at best. (If the cert contains multiple nodeIds as SAN, this logic can be broken, but I believe that is a corner-case that we don't need to support)
- If none of SAN matches with any nodeId, then use hostname verifier to support wildcard cert. In this case, return hostname derived from the socket address

Alternatively, we just need to use the hostname verifier and use the hostname derived from the socket address in any case for provenance data. How do you think @markap14 ?

markap14 replied:
I think falling back to the hostname from the socket whenever there is not an exact match (i.e., the wildcard matches but not an exact string comparison) is fair. Originally, we used the hostname directly from the socket, but as Koji mentioned in #3109 we changed that behavior. This was done because when you look at Provenance data (and in logs), what you may see is something like a RECEIVE event with a transit URI of nifi://s7302.r720.y8302.mydomain.com because that's the FQDN but the user typically references this node as say nifi-01.mydomain.com. If the UI shows the node as nifi-01.mydomain.com in the cluster table, then it is best to show that in the Provenance and logs as well. This is especially true in virtual environments, running in Docker or in a publish Cloud/VM where often the hostname reported by socket.getInetAddress() is very different than what we typically like to see.

Does that make sense?

Also @kotarot thank you for noticing this and submitting this contribution!

kotarot commented:
@markap14 Thank you for your kind advice! That makes sense to me.

In the new commit, I have left the existing authorization codes, followed by the authorization using `HostnameVerifier` (which I added). The authorization is performed as follows:
- If the authorization with the string-match succeeded, then returns the node information from Client Identities (as we fixed in #3109 ).
- After that, if the authorization by `HostnameVerifier` succeeded, then returns the derived hostname from the socket is returned.

Does this change seem to be no problem?

Also, I modified a few tests related to `LoadBalanceAuthorizer` because the interface of `authorize` is changed.

@ijokarumawak Could you please review it?

ijokarumawak commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/queue/clustered/server/ClusterLoadBalanceAuthorizer.java:
```diff
@@ -57,11 +79,35 @@ public String authorize(final Collection<String> clientIdentities) throws NotAut
+            }
+        }
+
+        final String message = String.format("Authorization failed for Client ID's to Load Balance data because none of the ID's are known Cluster Node Identifiers");
```
We don't have to use `String.format()` here, please the String to `logger.warn()` directly.

kotarot replied:
Thanks for pointing it out. I fixed it by just removing `String.format` in this line because the next line also uses the `message` variable.

ijokarumawak commented on nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/queue/clustered/server/ClusterLoadBalanceAuthorizer.java:
```diff
@@ -33,14 +42,27 @@
+
+        logger.debug("Will perform authorization against Client Identities '{}'", clientIdentities);
+
         if (clientIdentities == null) {
```
Now we only call this `authorize()` method if socket is a SSLSocket. We can remove this block.

kotarot replied:
Do you mean the block L66-69? Do we always guarantee `clientIdentities` is not null if the socket is a SSLSocket? I suppose we still need this.

ijokarumawak replied:
The existing log message indicating that this block is meant for the case where NiFi clustering is not secured (not sending data via SSLSocket). This block contradicts with the other block such as following getCertificateIdentities method:
```
        final Certificate[] certs = sslSession.getPeerCertificates();
        if (certs == null || certs.length == 0) {
            throw new SSLPeerUnverifiedException("No certificates found");
        }

```

If we care about `clientIdentities` being null, then we should throw `SSLPeerUnverifiedException("No client identities found");` instead of authorizing it. Having said that, I believe removing this block is safe as `clientIdentities` are populated using `Collectors.toSet`. If no SAN is found, the value will be an empty set instead of null.

kotarot replied:
@ijokarumawak OK, I get it now. Thanks for kindly telling me that. I pushed a new commit. Please check it. Thanks!

kotarot commented:
@ijokarumawak Thank you again for your review. I fixed the PR based on your comments. Can you check it?

ijokarumawak commented:
It looks good. +1. Merging. Thanks, @kotarot!

kotarot commented:
@ijokarumawak Thanks for reviewing and merging my PR!
